opnode/rollup/sync/start.go

```diff
@@ -59,45 +58,71 @@ var TooDeepReorgErr = errors.New("reorg is too deep")
 
 const MaxReorgDepth = 500
 
-// FindSafeL2Head takes the supplied L2 start block and walks the L2 chain until it finds the first L2 block reachable from the supplied
-// block that is also canonical.
-func FindSafeL2Head(ctx context.Context, start eth.BlockID, l1 L1Chain, l2 L2Chain, genesis *rollup.Genesis) (eth.L2BlockRef, error) {
-	// Starting point
-	l2Head, err := l2.L2BlockRefByHash(ctx, start.Hash)
+func isCanonical(ctx context.Context, l1 L1Chain, block eth.BlockID) (bool, error) {
+	canonical, err := l1.L1BlockRefByNumber(ctx, block.Number)
 	if err != nil {
-		return eth.L2BlockRef{}, fmt.Errorf("failed to fetch L2 head: %w", err)
+		return false, err
 	}
+	return canonical.Hash == block.Hash, nil
+}
+
+// FindL2Heads walks back from the supplied L2 blocks and finds the unsafe and safe L2 heads.
+// Unsafe Head: The first L2 block who's L1 Origin is canonical (determined via block by number).
+// Safe Head: Walk back 1 sequence window of epochs from the Unsafe Head.
+func FindL2Heads(ctx context.Context, start eth.L2BlockRef, seqWindowSize int,
+	l1 L1Chain, l2 L2Chain, genesis *rollup.Genesis) (unsafeHead eth.L2BlockRef, safeHead eth.L2BlockRef, err error) {
 	reorgDepth := 0
-	// Walk L2 chain from L2 head to first L2 block which has a L1 Parent that is canonical. May walk to L2 genesis
-	for n := l2Head; ; {
-		l1header, err := l1.L1BlockRefByNumber(ctx, n.L1Origin.Number)
-		if err != nil {
-			// Generic error, bail out.
-			if !errors.Is(err, ethereum.NotFound) {
-				return eth.L2BlockRef{}, fmt.Errorf("failed to fetch L1 block %v: %w", n.L1Origin.Number, err)
-			}
-			// L1 block not found, keep walking chain
-		} else {
-			// L1 Block found, check if matches & should keep walking the chain
-			if l1header.Hash == n.L1Origin.Hash {
-				return n, nil
+	var prevL1OriginHash common.Hash
+	// Walk L2 chain from start until L1Origin matches l1Base. Bail out early of when at genesis.
+	for n := start; ; {
+		// Check if l1Origin is canonical when we get to a new epoch
+		if prevL1OriginHash != n.L1Origin.Hash {
+			if ok, err := isCanonical(ctx, l1, n.L1Origin); err != nil {
+				return eth.L2BlockRef{}, eth.L2BlockRef{}, err
+			} else if ok {
+				unsafeHead = n
+				break
```

Comment:
Scenario:
- L2 chain is `abcde`
- L1 chain is `ABCDE`
- `start` is defined as the absolute L2 tip: `e`, an unsafe block derived from `D` as origin and included in `E`, already synced
- the local view of the L1 chain is only at `C`: so retrieval of `D` and `E` return `ethereum.NotFound`

In this case I would want to see `unsafeHead = e`, and `safeHead = c`.

But if we do what this PR currently does, then the `isCanonical` fails because `n.L1Origin = C` and `isCanonical` fails or returns false.

Instead, I think we should just remember the `l2UnsafeHead`, and only return something different if we find that the tip of L1 is not part of the any of the L1-origins that we traversed when walking back the L2 chain to the common L1 point.

This way we don't rewind the L2 chain suddenly:
- on restarts and such
- when we receive L1 blocks with a delay and don't do the extension-shortcut
- when we call the `findL2Heads` on e.g. receiving a new L2 block from an unsafe channel that is not a simple extension, but that doesn't reorg all unsafe blocks either.

Ideally we:
- call this function at any time to get the true sync starting point
- can stay on the latest L2 chain, even with reorgs, even if e.g. a L1 node is temporarily offline.

Reply:
I went through and updated this with a new test for it.

> on restarts and such

I think that we need to reorg the safe L2 chain on restarts because we don't track the end of the sequence window in way that can be checked at a later point.

> when we receive L1 blocks with a delay and don't do the extension-shortcut

This is case really only matters when we miss an L1 block. What happens is that the safe head / unsafe head should no-op change.

> when we call the findL2Heads on e.g. receiving a new L2 block from an unsafe channel that is not a simple extension, but that doesn't reorg all unsafe blocks either.

I'm not sure the use case here